crates/driver/src/api/mod.rs

```diff
@@ -1,42 +1,56 @@
-pub mod execute;
-pub mod solve;
-
 use {
-    crate::driver::Driver,
+    crate::solver::Solver,
     futures::Future,
-    shared::api::finalize_router,
+    nonempty::NonEmpty,
     std::{net::SocketAddr, sync::Arc},
-    tokio::{task, task::JoinHandle},
-    warp::{Filter, Rejection, Reply},
 };
 
-pub fn serve_api(
-    address: SocketAddr,
-    shutdown_receiver: impl Future<Output = ()> + Send + 'static,
-    drivers: Vec<(Arc<Driver>, String)>,
-) -> JoinHandle<()> {
-    let filter = handle_all_routes(drivers).boxed();
-    tracing::info!(%address, "serving driver");
-    let (_, server) = warp::serve(filter).bind_with_graceful_shutdown(address, shutdown_receiver);
-    task::spawn(server)
+pub mod execute;
+pub mod info;
+pub mod solve;
+
+const REQUEST_BODY_LIMIT: usize = 32 * 1024;
+
+pub async fn serve(
+    addr: &SocketAddr,
+    shutdown: impl Future<Output = ()> + Send + 'static,
+    solvers: NonEmpty<Solver>,
+) -> Result<(), hyper::Error> {
+    // Add middleware.
+    let app = axum::Router::new().layer(
+        tower::ServiceBuilder::new()
+            .layer(tower_http::limit::RequestBodyLimitLayer::new(
+                REQUEST_BODY_LIMIT,
+            ))
+            .layer(tower_http::trace::TraceLayer::new_for_http()),
+    );
+
+    // Add routes.
+    let app = solve::route(app);
+    let app = info::route(app);
```

Comment:
This is slightly different than what I had imagined and realised I didn't notice it when reading the issue https://github.com/cowprotocol/services/issues/895#issuecomment-1337498190.

I would change this a little. Basically, I would `solve::route(app, solver_name)` to mount one endpoint per solver in order to avoid doing competition in the `driver`. See the comment in the issue for more justification.